src/export/FSHExporter.ts

```diff
@@ -63,150 +75,190 @@ export class FSHExporter {
+    // for that profile. Otherwise it is written to instances.fsh
+    nonInlineInstances.forEach(instance => {
       if (instance.usage === 'Example' && files.has(`${instance.instanceOf}.fsh`)) {
-        files.get(`${instance.instanceOf}.fsh`).push(instance.toFSH());
-      } else if (files.has('instances.fsh')) {
-        files.get('instances.fsh').push(instance.toFSH());
+        files.get(`${instance.instanceOf}.fsh`).push(instance);
       } else {
-        files.set('instances.fsh', [instance.toFSH()]);
+        files.get('instances.fsh').push(instance);
       }
     });
+    // Inline instances are written to the file they are used in, if they are only used
+    // in one spot. Otherwise they go to instances.fsh
+    inlineInstances.forEach(instance => {
+      const usedIn = this.inlineInstanceUsedIn(instance, files);
+      usedIn.length === 1
+        ? files.get(usedIn[0]).push(instance)
+        : files.get('instances.fsh').push(instance);
+    });
+
+    // Invariants are written to the same file as the profile they are used in, if they
+    // are written in one spot. Otherwise they go to invariants.fsh.
+    files.set('invariants.fsh', []);
+    this.fshPackage.invariants.forEach(invariant => {
+      const usedIn = this.invariantUsedIn(invariant, files);
+      usedIn.length === 1
+        ? files.get(usedIn[0]).push(invariant)
+        : files.get('invariants.fsh').push(invariant);
+    });
 
     // All other artifacts are grouped by category
-    files.set('aliases.fsh', [this.fshPackage.aliases.map(a => a.toFSH()).join(EOL)]);
-    files.set(
-      'extensions.fsh',
-      this.fshPackage.extensions.map(extension => extension.toFSH())
-    );
-    files.set('terminology.fsh', [
-      ...this.fshPackage.valueSets.map(valueSet => valueSet.toFSH()),
-      ...this.fshPackage.codeSystems.map(codeSystem => codeSystem.toFSH())
-    ]);
-
-    files.set(
-      'invariants.fsh',
-      this.fshPackage.invariants.map(invariant => invariant.toFSH())
-    );
-    files.set(
-      'mappings.fsh',
-      this.fshPackage.mappings.map(mapping => mapping.toFSH())
-    );
+    files.set('aliases.fsh', this.fshPackage.aliases);
+    files.set('extensions.fsh', this.fshPackage.extensions);
+    files.set('valueSets.fsh', this.fshPackage.valueSets);
+    files.set('codeSystems.fsh', this.fshPackage.codeSystems);
+    files.set('mappings.fsh', this.fshPackage.mappings);
+
+    return files;
+  }
 
-    const exports: Map<string, string> = new Map();
-    files.forEach((content, fileName) => exports.set(fileName, content.join(`${EOL}${EOL}`)));
-    return exports;
+  private inlineInstanceUsedIn(
+    inlineInstance: ExportableInstance,
+    files: Map<string, Exportable[]>
+  ): string[] {
+    const usedIn: string[] = [];
+    files.forEach((exportables, file) => {
+      exportables
+        .filter(exportable => exportable instanceof ExportableInstance)
```

Comment:
I might be missing something here, but why do you filter to only look at other Instance? Couldn't inline instances be used on profiles as well?

Reply:
Yes they could, will fix